feat: Monty model parameter-counting script

nielsleadholm opened the pull request:
Adds a simple script to calculate the number of parameters of a Monty model, supporting one of the statements in the paper.

Assumptions etc. should all be clear in the doc string, but let me know if I can clarify anything. 

This might eventually be a useful mini-tutorial or the like, as I think it gives a helpful intuition of what a Monty model actually "is".

scottcanoe commented on scripts/count_monty_params.py:
```diff
@@ -0,0 +1,144 @@
+LM_ID = 0
+
+def count_monty_model_params(graph_memory: dict) -> int:
+    """
+    Count the number of parameters in a Monty model.
+
+    Args:
+        graph_memory (dict): The graph memory dictionary of the model.
+
+    Returns:
+        The total number of parameters in the model.
+    """
```
```suggestion
    """Count the number of parameters in a Monty model.

    Args:
        graph_memory (dict): The graph memory dictionary of the model.

    Returns:
        The total number of parameters in the model.
    """
```

scottcanoe commented on scripts/count_monty_params.py:
```diff
@@ -0,0 +1,144 @@
+)
+"""
+
+import os
+from pathlib import Path
+
+import torch
+
+# Get the value of the DMC_ROOT_DIR environment variable, or use "~/tbp/results/dmc" as
+# the default if not set
+DMC_ROOT_DIR = Path(os.environ.get("DMC_ROOT_DIR", "~/tbp/results/dmc")).expanduser()
```
```suggestion
import torch

# Get the value of the DMC_ROOT_DIR environment variable, or use "~/tbp/results/dmc" as
# the default if not set.
from data_utils import DMC_ROOT_DIR

```

scottcanoe commented on scripts/count_monty_params.py:
```diff
@@ -0,0 +1,144 @@
+# the default if not set
+DMC_ROOT_DIR = Path(os.environ.get("DMC_ROOT_DIR", "~/tbp/results/dmc")).expanduser()
+# We use the single-LM, distant-agent model.
+SINGLE_LM_MODEL_PATH = (DMC_ROOT_DIR /
+                        "pretrained_models/dist_agent_1lm/pretrained/model.pt")
```
```suggestion
DIST_AGENT_1LM_MODEL_PATH = (DMC_ROOT_DIR /
                        "pretrained_models/dist_agent_1lm/pretrained/model.pt")
```

scottcanoe commented on scripts/count_monty_params.py:
```diff
@@ -0,0 +1,144 @@
+
+
+if __name__ == "__main__":
+    graph_memory = torch.load(SINGLE_LM_MODEL_PATH)["lm_dict"][LM_ID]["graph_memory"]
```
```suggestion
    graph_memory = torch.load(DIST_AGENT_1LM_MODEL_PATH)["lm_dict"][LM_ID]["graph_memory"]
```

scottcanoe commented on scripts/count_monty_params.py:
```diff
@@ -0,0 +1,144 @@
+    graph_memory = torch.load(SINGLE_LM_MODEL_PATH)["lm_dict"][LM_ID]["graph_memory"]
+
+    total_params = count_monty_model_params(graph_memory)
+
+    print(f"Total number of parameters: {total_params}")
+    print(f"\nIn millions: {total_params / 1e6:.0f}M")
```
```suggestion

    print("Monty Model Parameter Count")
    print(f" - Model: {DIST_AGENT_1LM_MODEL_PATH}")
    print(f" - Sensory Channel: '{SENSORY_CHANNEL}'")
    print(f" - LM ID: {LM_ID}")
    print(f" - Total number of parameters: {total_params} (~{total_params / 1e6:.0f}M)")
```

scottcanoe commented on scripts/count_monty_params.py:
```diff
@@ -0,0 +1,144 @@
+# license that can be found in the LICENSE file or at
+# https://opensource.org/licenses/MIT.
+
+"""This script counts the number of parameters in a Monty model, specifically the
+dist_agent_1lm model. This is a result (4 million parameters) referenced in the
+Continual Learning section of the paper.
```
```suggestion
"""Count the number of parameters in the `dist_agent_1lm`  Monty model.

This script counts the number of parameters in a Monty model, specifically the
`dist_agent_1lm` pretrained model. This produces the result (4 million parameters) referenced in the
Continual Learning section of the paper.
```

scottcanoe commented on scripts/count_monty_params.py:
```diff
@@ -0,0 +1,144 @@
+dist_agent_1lm model. This is a result (4 million parameters) referenced in the
+Continual Learning section of the paper.
+
+The Monty model consists of a series of learned graphs for different objects. These
```
**praise**: Excellent description.

scottcanoe reviewed (approved):
Nice! I tried this PR, and works as expected on my machine.

I just left some optional (bikeshed) suggestions. They're basically just some renaming, docstring, and printing suggestions to be extra upfront the hard-coded model in question.

**thought**: I'd also be perfectly fine just doing the logic of `count_monty_model_params` inside` __main__`. It feels slightly awkward having `count_monty_model_params` use the global value `SENSORY_CHANNEL` rather than having it as a parameter, but it's not a big deal.

nielsleadholm commented:
Nice, thanks for all the useful feedback, looks a lot cleaner now. Also good catch re. passing the global variable. I've updated that and also added handling of multiple sensory channels - personally I like while implementing this to just spend 3 extra minutes so that it's more like a function ready to go and e.g. quickly re-discover and copy into `tbp.monty` when this use-case comes up again.

vkakerbeck commented on scripts/count_monty_params.py:
```diff
@@ -0,0 +1,165 @@
+- 'target_to_graph_id'
+- 'graph_id_to_target'
+The latter two are mappings between the object name and the index of the graph, which
+are needed to assess the performance of the model. They are of negligible size.
```
```suggestion
are needed to assess the performance of the model. 
```
I don't think you need to mention the size here as they are just used for logging and are not part of the model

vkakerbeck commented on scripts/count_monty_params.py:
```diff
@@ -0,0 +1,165 @@
+The accessed graph is a torch_geometric Data object, which has the following structure:
+
+Data(
+    x=[1188, 28],  # 28 features per node; not all of these are used in inference,
```
How many of them do we use? Should we just store the features that we actually use?

nielsleadholm replied:
That's determined by what we specify tolerances for right? I guess just these two then:

```
            tolerances={
                "patch": {
                    "hsv": np.array([0.1, 0.2, 0.2]),
                    "principal_curvatures_log": np.ones(2),
                }
            },
```

From:
https://github.com/thousandbrainsproject/tbp.tbs_sensorimotor_intelligence/blob/58f37a15ff5857858e92fef3d1fb50ccb1011bcc/monty/configs/common.py#L103

vkakerbeck replied:
ok so that would be 3+2+9=14 out of 28 that we use. (9 for the pose features) That would cut down the overall number of model parameters quite a bit, right? Seems reasonable to account for that.

nielsleadholm replied:
Yeah although I guess we also have to use the `pose_fully_defined`, `node_ids` etc. I might just leave it for now if it's ok but yeah we can update it if we want to emphasize the point even more (it's already a difference of 4M parameters for our model vs. 86 million for the smallest ViT).

vkakerbeck reviewed (approved):
Very nice and well documented! Agree that this could be something to add to the documentation
